### PR TITLE
[4.0] Move the additional language strings in a separate file & fix postinstall

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue https://github.com/joomla/joomla-cms/pull/18301#issuecomment-360113436

### Summary of Changes

> Is there a reason why all the language strings are in the .sys.ini language file and not spread across two language files as normal

Thanks @brianteeman 

- Fixing the postinstall behavior as this was broken before.

### Testing Instructions

- install this branch https://github.com/zero-24/joomla-cms/archive/httpheader.zip
- Check the config page of the httpheader plugin and make sure all language strings are still loaded. -
- Check the postinstall message and make sure all language strings are still loaded too. (please make sure to disable the plugin inorder to trigger the postinstall message)

### Expected result

two files & postinstall works also on new installs

### Actual result

one file & postinstall is broken.

### CODE REVIEW

**Comment by infograf768** [comment]
The xml file has no manifest part for languages. I suggest to add them.
```
	<languages>
		<language tag="en-GB">en-GB.plg_system_httpheader.ini</language>
		<language tag="en-GB">en-GB.plg_system_httpheader.sys.ini</language>
	</languages>
```

**Comment by infograf768** [comment]
There are other things that can be done for `customhttpheader.xml` but it could be done in another PR;
1. windows eol should be modified to Unix
2. No use to utilize specific strings for Site and Administrator as we already have the global ones, `JSITE` and `JADMINISTRATOR`.
3. String `PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_CLIENT="Site selection"` proposing Site, Administrator or Both, makes no sense. It should have "Client" as value.

**Comment by zero-24** [comment]
Thanks fixed.

File: administrator/components/com_admin/script.php [hunk 1]
```diff
@@ -166,6 +166,92 @@ protected function updateDatabase()
...
+				echo JText::sprintf('JLIB_DATABASE_ERROR_FUNCTION_FAILED', $e->getCode(), $e->getMessage()) . '<br>';
+
+				return;
+			}
+
+			// Save the params to the database
+			$query = $db->getQuery(true)
+				->update($db->quoteName('#__extensions'))
+				->set($db->quoteName('params') . ' = ' . $db->quote($params))
+				->where($db->quoteName('extension_id') . ' = 487')
```
Comment (Quy) [inline administrator/components/com_admin/script.php]: Should this be `700`?

Comment (infograf768) [reply administrator/components/com_admin/script.php]: As far as I see 700 is a common id in postinstall table, 487 in the extensions table

Comment (zero-24) [reply administrator/components/com_admin/script.php]: sorry it looks like i missed your messages :(

I have now choose a different way to solve this problem by renaming the core plugin it is not going to conflict with the 3.x version on a upgrade. So no need for that complete sql stuff.

**Comment by zero-24** [comment]
@wilsonge if you have a few minutes please take care that this is merged before the next release as this is going to rename the current plg_system_httpheader to plg_system_httpheaders so it is not going to conflict with the original version currently used in 3.x

File: plugins/system/httpheaders/postinstall/introduction.php [hunk 2]
```diff
@@ -1,33 +1,33 @@
...
  *
  * @return  integer
  *
  * @since   4.0
  */
-function httpheader_postinstall_condition()
+function httpheaders_postinstall_condition()
 {
-	return Joomla\CMS\Plugin\PluginHelper::isEnabled('system', 'httpheader');
+	return !Joomla\CMS\Plugin\PluginHelper::isEnabled('system', 'httpheader');
```
Comment (Quy) [inline plugins/system/httpheaders/postinstall/introduction.php]: Here too?

Comment (zero-24) [reply plugins/system/httpheaders/postinstall/introduction.php]: just pushed / fixed thanks 👍

**Comment by infograf768** [comment]
I was not aware we had a plg_system_httpheader  in 3.x

**Comment by zero-24** [comment]
> I was not aware we had a plg_system_httpheader in 3.x

https://github.com/zero-24/plg_system_httpheader

This is the base version for the 4.0.0 and you can use that on any 3.xer site you want :)

File: plugins/system/httpheaders/httpheaders.xml [hunk 3]
```diff
@@ -49,28 +49,32 @@
...
 				</field>
 				<field
-					name="additional_httpheader"
+					name="additional_httpheaders"
 					type="subform"
-					label="PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER"
-					description="PLG_SYSTEM_HTTPHEADER_ADDITIONAL_HEADER_DESC"
-					formsource="plugins/system/httpheader/subform/customhttpheader.xml"
+					label="PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER"
+					description="PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_DESC"
```
Comment (Quy) [inline plugins/system/httpheaders/httpheaders.xml:69]: Description is displayed in the tooltip and as text. OK?

httpheader-tooltip

**Comment by Quy** [comment]
`Referrer-Policy` setting will not save `Disabled`.

File: plugins/system/httpheaders/httpheaders.xml (hunk 3 above)
Comment (zero-24) [reply plugins/system/httpheaders/httpheaders.xml:69]: This is how 4.0 currently render the tooltips as most of them are dropped I have no control over that.

Comment (brianteeman) [reply plugins/system/httpheaders/httpheaders.xml:69]: Still waiting a decision from the a11y folks on one of the proposed solutions for that.

Comment (zero-24) [reply plugins/system/httpheaders/httpheaders.xml:69]: Thanks! 4.0 is under development so everything is great. 👍  ;)

**Comment by zero-24** [comment]
Just pushed the changes for disabled.

File: plugins/system/httpheaders/httpheaders.php [hunk 4]
```diff
@@ -124,6 +124,11 @@ private function setDefaultHeader()
 		}
 
 		// Referrer-Policy
-		$this->app->setHeader('Referrer-Policy', $this->params->get('referrerpolicy', 'no-referrer-when-downgrade'));
+		$referrerpolicy = $this->params->get('referrerpolicy', 'no-referrer-when-downgrade');
+
+		if ($referrerpolicy != 'disabled')
```
Comment (Quy) [inline plugins/system/httpheaders/httpheaders.php]: `!==`

**Comment by Quy** [comment]
Remove extra blank lines.

The `Enable default security headers` button does nothing.

header-postinstall

**Comment by zero-24** [comment]
> The Enable default security headers button does nothing.

can you please double check that? It should just enable the plugin currently there is no redirection implemented.

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
...><li><a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-frame-options'>'X-Frame-Options: SAMEORIGIN'</a></li><li><a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-xss-protection'>'X-XSS-Protection: 1; mode=block'</a></li><li><a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-content-type-options'>'X-Content-Type-Options: nosniff'</a></li><li><a href='https://scotthelme.co.uk/a-new-security-header-referrer-policy/'>'Referrer-Policy: no-referrer-when-downgrade'</a></li></ul><br>The full list of supported headers are: <br><ul><li><a href='https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security'>Strict-Transport-Security</a></li><li><a href='https://en.wikipedia.org/wiki/Content_Security_Policy'>Content-Security-Policy</a></li><li>Content-Security-Policy-Report-Only</li><li>X-Frame-Options</li><li>X-XSS-Protection</li><li>X-Content-Type-Options</li><li>Referrer-Policy</li><li>Expect-CT</li></ul><br>These headers help your browser to protect your website from <a href='https://en.wikipedia.org/wiki/Cross-site_scripting'>XSS</a> and <a href='https://en.wikipedia.org/wiki/Clickjacking'>Clickjacking</a> attacks.</p>"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: `<ul>` tag should not be inside a `<p>` tag.

**Comment by Quy** [comment]
Still disabled. I expect the postinstall message to disappear after clicking the button if the plugin is to be enabled.

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini (hunk 5 above)
Comment (zero-24) [reply administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: `<p>` - tag is removed now.

**Comment by zero-24** [comment]
> Still disabled. I expect the postinstall message to disappear after clicking the button if the plugin is enabled.

Sure. hmm I'm going to debug this one. Maybe i'm also going to redirect the user to the edit page so they can configure the needed values in the plugin if they hit enable. Let's see what is the problem and how to fix it.

**Comment by zero-24** [comment]
I have just fixed the button. A redirect to the plugin edit page is not possible ;)

**Comment by brianteeman** [comment]
Yes it is - i did it in the recaptcha v1 pr #19648  or i misunderstand

**Comment by zero-24** [comment]
I'm going to review that proposal. Thanks!

**Comment by zero-24** [comment]
hmm look like this is working. My test before that resulted into a "you are not allowed to use this link directly" error. Now this is fixed. So commit is in coming. Thanks @brianteeman

**Comment by Quy** [comment]
The default headers are not set. A var dump of `$this->params->get('xframeoptions')` results in `string(1) "1"` which is not the same type as `===1`.

```
		// X-Frame-Options
		if ($this->params->get('xframeoptions', 1) === 1)
		{
			$this->app->setHeader('X-Frame-Options', 'SAMEORIGIN');
		}
```

**Comment by zero-24** [comment]
Thanks should be fixed by the last commit.

**Comment by Quy** [comment]
I had to save the plugin initially before it would be in effect. `1` should be a string as follows:

`get('xframeoptions', '1')`

**Comment by Quy** [comment]
I am testing with your branch rather than via PatchTester. This is the error I get when doing a fresh install. I know it is unrelated to your PR, but how to fix it?
`0 Could not create session directory "/var/lib/php/session"`

**Comment by zero-24** [comment]
Hmm i have no idea how the session things works nowdays in 4.0 can your user write to that path?

**Comment by wilsonge** [comment]
Thanks Tobias :)

**Comment by zero-24** [comment]
🎉

**Pull request merged**